titus-server-master/src/main/java/com/netflix/titus/master/scheduler/constraint/KubeConstraint.java

```diff
@@ -0,0 +1,154 @@
+ */
+@Singleton
+public class KubeConstraint implements SystemConstraint {
+
+    public static final String NAME = "KubeConstraint";
+    public static final String INSTANCE_NOT_FOUND_REASON = "Instance for node not found";
+    public static final String NOT_NOT_READY_REASON = "Node's ready node condition is not true";
+    public static final String TAINT_NOT_TOLERATED_IN_CONFIGURATION_REASON = "Node has a taint that is not configured to be tolerated";
+    public static final String READY = "Ready";
+
+    private static final Result VALID = new Result(true, null);
+
+    private enum Failure {
+        INSTANCE_NOT_FOUND(INSTANCE_NOT_FOUND_REASON),
+        NODE_NOT_READY(NOT_NOT_READY_REASON),
+        TAINT_NOT_TOLERATED_IN_CONFIGURATION(TAINT_NOT_TOLERATED_IN_CONFIGURATION_REASON);
+
+        private final Result result;
+
+        Failure(String reason) {
+            this.result = new Result(false, reason);
+        }
+
+        public Result toResult() {
+            return result;
+        }
+    }
+
+    private static final Set<String> FAILURE_REASONS = Stream.of(KubeConstraint.Failure.values())
+            .map(f -> f.toResult().getFailureReason())
+            .collect(Collectors.toSet());
+
+    public static boolean isKubeConstraintReason(String reason) {
+        return reason != null && FAILURE_REASONS.contains(reason);
+    }
+
+    private final SchedulerConfiguration schedulerConfiguration;
+    private final MesosConfiguration mesosConfiguration;
+    private final AgentManagementService agentManagementService;
+    private final KubeApiFacade kubeApiFacade;
+
+
+    @Inject
+    public KubeConstraint(SchedulerConfiguration schedulerConfiguration,
+                          MesosConfiguration mesosConfiguration,
+                          AgentManagementService agentManagementService,
+                          KubeApiFacade kubeApiFacade) {
+        this.schedulerConfiguration = schedulerConfiguration;
+        this.mesosConfiguration = mesosConfiguration;
+        this.agentManagementService = agentManagementService;
+        this.kubeApiFacade = kubeApiFacade;
+    }
+
+    @Override
+    public String getName() {
+        return NAME;
+    }
+
+    @Override
+    public void prepare() {
+    }
+
+    @Override
+    public Result evaluate(TaskRequest taskRequest, VirtualMachineCurrentState targetVM, TaskTrackerState taskTrackerState) {
+        if (!mesosConfiguration.isKubeApiServerIntegrationEnabled()) {
+            return VALID;
+        }
+
+        Optional<AgentInstance> instanceOpt = SchedulerUtils.findInstance(agentManagementService, schedulerConfiguration.getInstanceAttributeName(), targetVM);
+        if (!instanceOpt.isPresent()) {
+            return Failure.INSTANCE_NOT_FOUND.toResult();
+        }
+
+        String instanceId = instanceOpt.get().getId();
+        V1Node node = kubeApiFacade.getNodeInformer().getIndexer().getByKey(instanceId);
```

Comment:
What if node is not there? We should check for null at least + double check that an exception is not thrown if a node is missing or due to other problems.

Reply:
good catch. updated